### PR TITLE
feat: node types

### PR DESCRIPTION
**Pull request by paolobrasolin**

### CODE REVIEW

**Review by jonsterling — COMMENTED** [review]
This looks like just what I was hoping for!

**Review by jonsterling — COMMENTED** [review]
Do you think this feature is ready to be merged, or are there things that you want to get done before merging?

**Comment by paolobrasolin** [comment]
> Do you think this feature is ready to be merged, or are there things that you want to get done before merging?

I think this is still too rough. I'll be able to work on this tomorrow, if you can wait @jonsterling :smile:

**Comment by jonsterling** [comment]
Haha take your time! :)

**Comment by paolobrasolin** [comment]
Ok @jonsterling I'm feeling good about this!
What's a theorem/lemma you'd like to detach as an independent node, so I can commit a usage demo to this PR?

**Comment by jonsterling** [comment]
Sick!!! Let's try splitting http://www.jonmsterling.com/math/nodes/000O.html into a definition; the definition starts at the words "Let E be a category...".

**Comment by paolobrasolin** [comment]
@jonsterling done!

Any formatting detail should be controllable modifying the layouts.

**Comment by jonsterling** [comment]
Cool, I'm playing with it now!

Feature request: can we omit have the table of contents span only section nodes? (i.e. omit result nodes from it)

**Comment by jonsterling** [comment]
I'm not totally certain my request is a good idea BTW, because it might make the section numbering be unexpected.

**Comment by jonsterling** [comment]
Let me mention something that I did in my thesis which kind of solves the node numbering issue (though taste may differ). 

Basically it is nice for all nodes to have readable numbers, but conventionally only section nodes would appear in the TOC; if we want the TOC to not appear to skip numbers, we have a problem. My suggestion is that each section node should induce a flat namespace of terminal/result nodes within it. Concretely this could look like:

# 1. Section

## 1.1. Subsection

**Definition 1.1❋a.** Blah Blah

**Theorem 1.1❋b.** Blah blah blah

### 1.1.1. Subsubsection

**Comment by paolobrasolin** [comment]
> Feature request: can we omit have the table of contents span only section nodes? (i.e. omit result nodes from it)

Yes, that can be done at the template level in `toc.html`:

```
{% assign filtered_resources = include.page.children | where: "sublayout", "sheafy/node/section" -%}
{% for resource in filtered_resources %}
  ...
{% endfor %}
```

The reason this didn't even occur to me is that Stacks just lists everything in the TOC, e.g.: 
image

> I'm not totally certain my request is a good idea BTW, because it might make the section numbering be unexpected.

I'm not sure either, but I guess it could be boiled down to taste. Maybe labeling `result`s in the TOC as "Theorem"/"Definition"/etc (like Stacks above) would solve the confusion created when they're mixed with sections? (I'm guessing that's the reason you'd like them hidden in the first place, but I might be completely off track.)

Anyways, putting aside taste considerations, creating a separate namespace for the numbering on terminal nodes like you suggest is the only way to avoid incongruences when hiding them. Doing that involves either changing the mechanism which is generating the `iso_2145` or completely generating the numbering at the template level (definitely doable, a bit involved).

Numbering is definitely subject to taste, so `sheafy` should not tie our hands too much. I'll toy around a bit and see if I can come up with a clever solution. :thinking: 

Btw: a toggle to show/hide leaves in the TOC might also be interesting! :shrug:

**Comment by jonsterling** [comment]
Thanks! So my main reason for wanting them hidden is that they tend to break the flow of the textbook, since the TOC for me is about high-level organization of concepts and this is diluted when flooded by a sea of `Definition`, `Theorem`, etc.  As you can see, the deepest level of TOC in Stacks is not very informative. 

I am in principle open to them being visible, but I think it creates some friction. For the TOC to not be super ugly and cluttered, one needs a meaningful title for every definition and theorem. But in context, such titles are often highly redundant. `Definition (monoid closed category). A monoidal closed category is ...` For this reason, terminal result nodes feel very different from sections.

**Comment by jonsterling** [comment]
Another possibility is that we simply never allow a `result` node to be a sibling of a `section` node. That would also solve the problem... The Stacks Project obeys this invariant, for instance. It might require me to rethink the way I write mathematics a bit, but it is definitely something that we could consider.

**Comment by paolobrasolin** [comment]
> It might require me to rethink the way I write mathematics a bit

This alone seems reason enough to search for a better solution, IMO.

---

I gave this some though and I believe the Right:tm: thing to do is:
* abolish the distinctions of Jekyll `collection`s inside `sheafy` (it's been bugging me for a while: it makes things awkward)
* once the treatment of nodes is homogeneous, classify them into _taxa_ by inserting lines like `sheafy: section` in nodes frontmatters
* `sheafy` will then expand _taxa_ to a set of default key/value pairs defined in `_config.yml`

Here's the kicker: the default metadata for a _taxon_ also defines things like `toc` (a boolean to toggle its rendering in TOC), `counter` (the name of the counter to be used for numbering), default metadata for shared layouts (e.g. a "Theorem" label for the `theorem` taxon which shares its layout with others), ...

This is what the https://github.com/jonsterling/math/pull/28/commits/cfb1f7f114236d08a5d7ab70e26217ac3bcb23ea commit sketches. Note I fixed the TOC fitlering but not yet the numbering.

Once we have this simple way to set default metadata for any node in a _taxon_, the code basically writes itself!

We can have semantic properties for filtering, like in the TOC:

```
  {% assign filtered_resources = include.page.children | where: "toc", true -%}
```

Nodes metadata becomes much simpler:

```
---
title: Internal category
sheafy: definition
---
```

And here is the configuration:

```yaml
collections:
  lectures:
    output: true
  nodes:
    output: true

defaults:
  - scope:
      type: nodes
    values:
      sheafy: section # this just means: by default an item of collection "nodes" has taxon "section"
      excerpt: ""
  - scope:
      type: lectures
    values:
      sheafy: section # this just means: by default an item of collection "nodes" has taxon "section"
      excerpt: ""

# all configuration of collections is above this point and the user can do whathever she pleases with them
# all configuration of taxa lies below this point and the user has a proper way to reason about properties of "node types"

sheafy:
  taxa:
    section:
      toc: true
      layout: sheafy/tree/default
      sublayout: sheafy/node/section
    theorem:
      toc: false
      layout: sheafy/tree/default
      sublayout: sheafy/node/section
      genus: Theorem
    definition:
      toc: false
      layout: sheafy/tree/default
      sublayout: sheafy/node/section
      genus: Definition
```

I believe this strikes a good balance of generality and usability! What do you think @jonsterling ?
If you too feel that this is a good lead, I'll handle the numbering issues we talked about and clean this up.

**Comment by jonsterling** [comment]
Wow, this is very cool and looks really nice for customization :)

**Comment by paolobrasolin** [comment]
Ok @jonsterling I think this is ready:

* I removed the `toc` property and I'm using the taxa directly to filter nodes to generate the TOC (it makes more sense to me, but now that taxa exist you can define arbitrary node properties to use in layouts)
* I removed `iso_2146` and `numbering`. We're now using named `clicker`s (configured per-taxa in `_config.yml`) which act like "telescopic counters" on the nodes. Concretely, a node now has its `clicks` attached -- something like `[{"clicker"=>"section", "value"=>0}, {"clicker"=>"section", "value"=>5}, {"clicker"=>"section", "value"=>0}, {"clicker"=>"result", "value"=>0}]` -- which then gets rendered as `6.1❋a.` by a filter like so: `{{ resource.clicks | to_numbering }}`. The filter `to_numbering` is an easily customized plugin and completely controls how clicks are rendered to "human numbering".

Ok; there's a lot in this PR. Let's unpack:

* nodes rendered as standalone are wrapped in their `layout`
* nodes rendered by `@include` are wrapped in their `sublayout` (allowing theorem, definitions, etc)
* nodes are now identified (not by collections but) by the attribute `sheafy` in their frontmatter; its value is the node's `taxon`, which in turn gets expanded into the relative attributes configured in `_config.yml`
* numbering is now handled abstractly by named clickers which populate the `clicks` attribute on nodes, which can the be rendered however one desires by defining a custom filter
* TOC now shows only nodes of the `section` taxon

I feel pretty good about this design!

Oh, yeah, here's a screenshot:

image

Do you want to do a last test run?
I'll solve the conflicts then and prepare for merging.

File: _layouts/sheafy/tree/default.html
Comment (jonsterling) [inline _layouts/sheafy/tree/default.html]: I am noticing that the genus doesn't show up, and I think this might be the place that is causing it. Should this be its own conditional, prior to the title?

File: _config.yml
Comment (jonsterling) [inline _config.yml]: So what exactly does the `sheafy` field do?

File: _config.yml
Comment (jonsterling) [inline _config.yml:106]: One thing I would love is for the standalone layout of result nodes to basically look like their sublayout. I tried doing so myself with a custom layout but I got a bit turned around. Is this possible?

What I have in mind is kind of like how on the Stacks Project pages for lemmas, it still shows the `Lemma 000C. blah blah blah` instead of a big `<h1>`: https://stacks.math.columbia.edu/tag/000C

File: _config.yml
Comment (jonsterling) [reply _config.yml]: Oh, I see your explanation in the comment above. I think I get it now! Thanks.

File: _config.yml
Comment (jonsterling) [reply _config.yml:106]: Btw the thing I had trouble with when trying to do this myself was that I needed the `Definition foo. ` to get spliced into the first paragraph of the body of the node, but instead it is is broken off by a paragraph boundary. I didn't quite understand why this wasn't an issue when the node is being rendered within a section, but I figured you'd know what's going on.

File: _layouts/sheafy/tree/default.html
Comment (paolobrasolin) [reply _layouts/sheafy/tree/default.html]: Actually it is, but probably not in the way you'd expect -- i.e. the work is working as _I_ expect.

Definition with title, rendered as standalone:

image

Definition without title, rendered as standalone:

image

All this is to avoid showing ugly stuff like `6.1❋a 0015 [0015]` unless unavoidable, without being too verbose, as it's a huge heading.

This problem is related to https://github.com/jonsterling/math/pull/28#discussion_r793912032 so let's continue there.

File: _config.yml
Comment (paolobrasolin) [reply _config.yml:106]: Yes, it's possible. I tried and opted for the current solution just because it's a bit simpler. I'll take care of it now.
Btw -- you still want the sections to have an enlarged heading when they're rendered as roots, right?

---

**UPDATE**
Since only `section`s are meant to have a TOC, maybe I can just move it to the `section` sublayout.
And this probably makes the `topmost` attribute irrelevant. :thinking: 

---

**UPDATE**
Actually, this is a really good change as it allows me to write better semantic HTML!

Comment (jonsterling) [reply _config.yml:106]: !! cool! And yes, I think TOC only needs to appear on sections.

Thanks for being so patient with me! Your work is excellent and I'm very excited by it.

Comment (paolobrasolin) [reply _config.yml:106]: @jonsterling done with the layouts!

I also replaced the broken `topmost` attribute with the `subroot` one which contains the root of the subtree being rendered in context (needed to reliably render only its TOC in the sublayout).

**Review by jonsterling — APPROVED** [review]
I just tried this out and it works just like I'd hoped! Really nice work.

My last question is, should the `sheafy` field in node metadata be renamed to `taxon`? (I don't feel strongly about it, just wanted to ask!)

Feel free to merge whenever you feel it's ready.

**Comment by jonsterling** [comment]
Oh, one last thing I noticed. I think for section nodes it would be preferable that the level of the HTML `<h#>` tag on their standalone pages start at the top, regardless of the level of the node. (To be clear, this shouldn't block the merge if you prefer that we address that elsewhere.) **[I actually think we should get this PR merged right away, and discuss that thing later.]**

**Comment by paolobrasolin** [comment]
> My last question is, should the `sheafy` field in node metadata be renamed to `taxon`? (I don't feel strongly about it, just wanted to ask!)

Yeah, I had a bit of back and forth on that. `sheafy` avoids conflicts but `taxon` feels better, though. At the end it's simple to make this customizable, so I'll rename it to `taxon`. :+1: 

> I actually think we should get this PR merged right away, and discuss that thing later

Ok!

> Feel free to merge whenever you feel it's ready.

On we go! :rocket:

**Pull request merged**
